akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala

```diff
@@ -0,0 +1,377 @@
+      }
+    }
+
+    res
+  }
+
+  def applySideEffect(effect: ChainableEffect[_, S]): Behavior[Any] = effect match {
+    case _: Stop.type @unchecked ⇒
+      Behaviors.stopped
+
+    case SideEffect(callbacks) ⇒
+      callbacks(state)
+      same
+  }
+
+  def applyEvent(s: S, event: E): S =
+    eventHandler(s, event)
+
+  @tailrec private def applyEffects(msg: Any, effect: Effect[E, S], sideEffects: immutable.Seq[ChainableEffect[_, S]] = Nil): Behavior[Any] = {
+    log.info(s"ApplyEffects: ${msg} => Effect: ${effect}")
+
+    effect match {
+      case CompositeEffect(Some(persist), currentSideEffects) ⇒
+        applyEffects(msg, persist, currentSideEffects ++ sideEffects) // "unwrap" & recur
+
+      case CompositeEffect(_, currentSideEffects) ⇒
+        applySideEffects(currentSideEffects ++ sideEffects) unlessStopped tryUnstash(context, this)
+
+      case Persist(event) ⇒
+        // apply the event before persist so that validation exception is handled before persisting
+        // the invalid event, in case such validation is implemented in the event handler.
+        // also, ensure that there is an event handler for each single event
+        state = applyEvent(state, event)
+        val tags = tagger(event)
+        val eventToPersist = if (tags.isEmpty) event else Tagged(event, tags)
+
+        log.info(s" >>>> [${context.self.path.name}] in Persist($event) @ ${lastSequenceNr}")
+
+        internalPersist(eventToPersist) { _ ⇒
+          applySideEffects(sideEffects)
+
+          log.info(s" >>>> [${context.self.path.name}] in Persist($event) CALLBACK @ ${lastSequenceNr}")
+
+          log.info(s"persist callback [${context.self.path.name}] == seqNr=${lastSequenceNr} => ${snapshotWhen(state, event, lastSequenceNr)}")
+          if (snapshotWhen(state, event, lastSequenceNr))
+            internalSaveSnapshot(state)
+        }
+
+      case PersistAll(events) ⇒
+        if (events.nonEmpty) {
+          // apply the event before persist so that validation exception is handled before persisting
+          // the invalid event, in case such validation is implemented in the event handler.
+          // also, ensure that there is an event handler for each single event
+          var count = events.size
+          var seqNr = lastSequenceNr
+          val (newState, shouldSnapshotAfterPersist) = events.foldLeft((state, false)) {
+            case ((currentState, snapshot), event) ⇒
+              seqNr += 1
+              val shouldSnapshot = snapshot || snapshotWhen(currentState, event, seqNr)
+              (applyEvent(currentState, event), shouldSnapshot)
+          }
+          state = newState
+          val eventsToPersist = events.map { event ⇒
+            val tags = tagger(event)
+            if (tags.isEmpty) event else Tagged(event, tags)
+          }
+          internalPersistAll(eventsToPersist) { _ ⇒
+            count -= 1
+            if (count == 0) {
+              sideEffects.foreach(applySideEffect)
```

Comment:
returned behavior not used, and similar in more places

Reply:
fixed now